In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
pd.set_option('display.max_columns', 152)
pd.set_option('display.max_rows', 152)

In [2]:
df = pd.read_csv('accepted_2007_to_2018Q4.csv')
df.drop(columns='member_id',inplace=True)


/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3063: DtypeWarning: Columns (0,19,49,59,118,129,130,131,134,135,136,139,145,146,147) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
df.sample(8)

id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
1033852   73500327     6500.0       6500.0           6500.0   36 months   
1392022  144376173     5000.0       5000.0           5000.0   36 months   
907159   109517489    28250.0      28250.0          28250.0   36 months   
732670    77028940     8000.0       8000.0           8000.0   36 months   
648753   112148644     8000.0       8000.0           8000.0   36 months   
2027210   84789385    15000.0      15000.0          15000.0   36 months   
714570    76449374    20000.0      20000.0          20000.0   36 months   
2197961   93861232    12000.0      12000.0          12000.0   36 months   

         int_rate  installment grade sub_grade  \
1033852      6.49       199.19     A        A2   
1392022     12.98       168.43     B        B5   
907159      22.91      1092.23     E        E1   
732670       6.49       245.16     A        A2   
648753      11.99       265.68     B        B5   
2027210     11.49       494.57     B        B5   
714570       9.16       637.49     B        B2   
2197961      8.24       377.37     B        B1   

                                      emp_title emp_length home_ownership  \
1033852                        Registered nurse    3 years            OWN   
1392022                       Sr Ops Consultant  10+ years           RENT   
907159                       Equipment Operator  10+ years           RENT   
732670                         Registered Nurse    2 years            OWN   
648753                          Owner/President    9 years           RENT   
2027210                    Assistant Principal   10+ years       MORTGAGE   
714570   Clinical Resource Management Assistant    2 years       MORTGAGE   
2197961                      Service Technician   < 1 year           RENT   

         annual_inc verification_status   issue_d loan_status pymnt_plan  \
1033852     65000.0        Not Verified  Mar-2016  Fully Paid          n   
1392022     80000.0        Not Verified  Dec-2018     Current          n   
907159      85000.0            Verified  Jun-2017     Current          n   
732670      50000.0        Not Verified  Apr-2016     Current          n   
648753      51500.0            Verified  Jul-2017     Current          n   
2027210     98000.0     Source Verified  Jul-2016  Fully Paid          n   
714570      71000.0     Source Verified  May-2016     Current          n   
2197961     78500.0     Source Verified  Nov-2016  Fully Paid          n   

                                                       url desc  \
1033852  https://lendingclub.com/browse/loanDetail.acti...  NaN   
1392022  https://lendingclub.com/browse/loanDetail.acti...  NaN   
907159   https://lendingclub.com/browse/loanDetail.acti...  NaN   
732670   https://lendingclub.com/browse/loanDetail.acti...  NaN   
648753   https://lendingclub.com/browse/loanDetail.acti...  NaN   
2027210  https://lendingclub.com/browse/loanDetail.acti...  NaN   
714570   https://lendingclub.com/browse/loanDetail.acti...  NaN   
2197961  https://lendingclub.com/browse/loanDetail.acti...  NaN   

                    purpose                    title zip_code addr_state  \
1033852         credit_card  Credit card refinancing    550xx         MN   
1392022         credit_card  Credit card refinancing    337xx         FL   
907159   debt_consolidation       Debt consolidation    816xx         CO   
732670          credit_card  Credit card refinancing    648xx         MO   
648753   debt_consolidation       Debt consolidation    902xx         CA   
2027210  debt_consolidation       Debt consolidation    330xx         FL   
714570   debt_consolidation                      NaN    920xx         CA   
2197961  debt_consolidation       Debt consolidation    531xx         WI   

           dti  delinq_2yrs earliest_cr_line  fico_range_low  fico_range_high  \
1033852  18.52          0.0         Jan-2003           730.0            734.0   
1392022  39.70          0.0         Oct-1988           685.

In [4]:
df.dtypes

id                                             object
loan_amnt                                     float64
funded_amnt                                   float64
funded_amnt_inv                               float64
term                                           object
int_rate                                      float64
installment                                   float64
grade                                          object
sub_grade                                      object
emp_title                                      object
emp_length                                     object
home_ownership                                 object
annual_inc                                    float64
verification_status                            object
issue_d                                        object
loan_status                                    object
pymnt_plan                                     object
url                                            object
desc                        

In [5]:
df.shape

(2260701, 150)

In [6]:
df.columns

Index(['id', 'loan_amnt', 'funded_amnt', 'funded_amnt_inv', 'term', 'int_rate',
       'installment', 'grade', 'sub_grade', 'emp_title',
       ...
       'hardship_payoff_balance_amount', 'hardship_last_payment_amount',
       'disbursement_method', 'debt_settlement_flag',
       'debt_settlement_flag_date', 'settlement_status', 'settlement_date',
       'settlement_amount', 'settlement_percentage', 'settlement_term'],
      dtype='object', length=150)

In [7]:
df.isnull().sum()

id                                                  0
loan_amnt                                          33
funded_amnt                                        33
funded_amnt_inv                                    33
term                                               33
int_rate                                           33
installment                                        33
grade                                              33
sub_grade                                          33
emp_title                                      167002
emp_length                                     146940
home_ownership                                     33
annual_inc                                         37
verification_status                                33
issue_d                                            33
loan_status                                        33
pymnt_plan                                         33
url                                                33
desc                        

In [8]:
#df.drop(columns='member_id',inplace=True)

In [9]:
df = df[df['issue_d'].notna()]

In [10]:
df['year']=df.issue_d.map(lambda x:int(x.split('-')[-1]))
# Get names of indexes for which column Age has value 30
indexNames = df[ df['year'] <=2012 ].index
# Delete these row indexes from dataFrame
df.drop(indexNames , inplace=True)
df.shape

(2164766, 151)

In [11]:
df.sample(5)

id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
830947   136524448    11500.0      11500.0          11500.0   36 months   
2141900  117837309     8000.0       8000.0           8000.0   36 months   
1310396   13037351    10000.0      10000.0          10000.0   36 months   
792079   139082570     4800.0       4800.0           4800.0   36 months   
1804279    7094676    18000.0      18000.0          18000.0   36 months   

         int_rate  installment grade sub_grade                   emp_title  \
830947       6.11       350.43     A        A1                     Teacher   
2141900     17.09       285.59     D        D1                         NaN   
1310396     12.99       336.90     B        B5         Attorney/of counsel   
792079      19.92       178.19     D        D3  Universal Home Tech Expert   
1804279      8.90       571.56     A        A5    Cinti Country Day SChool   

        emp_length home_ownership  annual_inc verification_status   issue_d  \
830947   10+ years       MORTGAGE     50000.0            Verified  Aug-2018   
2141900        NaN            OWN     75000.0            Verified  Oct-2017   
1310396    2 years           RENT    150000.0     Source Verified  Mar-2014   
792079   10+ years       MORTGAGE     65000.0        Not Verified  Aug-2018   
1804279  10+ years       MORTGAGE     55000.0     Source Verified  Sep-2013   

         loan_status pymnt_plan  \
830947       Current          n   
2141900      Current          n   
1310396  Charged Off          n   
792079       Current          n   
1804279   Fully Paid          n   

                                                       url  \
830947   https://lendingclub.com/browse/loanDetail.acti...   
2141900  https://lendingclub.com/browse/loanDetail.acti...   
1310396  https://lendingclub.com/browse/loanDetail.acti...   
792079   https://lendingclub.com/browse/loanDetail.acti...   
1804279  https://lendingclub.com/browse/loanDetail.acti...   

                                                      desc  \
830947                                                 NaN   
2141900                                                NaN   
1310396                                                NaN   
792079                                                 NaN   
1804279    Borrower added on 08/29/13 > I am trying to ...   

                    purpose                    title zip_code addr_state  \
830947     home_improvement         Home improvement    356xx         AL   
2141900         credit_card  Credit card refinancing    153xx         PA   
1310396  debt_consolidation       Debt consolidation    806xx         CO   
792079     home_improvement         Home improvement    700xx         LA   
1804279  debt_consolidation       Debt consolidation    450xx         OH   

           dti  delinq_2yrs earliest_cr_line  fico_range_low  fico_range_high  \
830947    1.20          0.0         Sep-2004           765.0            769.0   
2141900  23.78          1.0         Dec-1989           700.0            704.0   
1310396  14.22          4.0         May-1989           660.0            664.0   
792079   13.33          0.0         Jul-2003           710.0            714.0   
1804279  24.49          0.0         Dec-1995           680.0            684.0   

         inq_last_6mths  mths_since_last_delinq  mths_since_last_record  \
830947              0.0                    30.0                     NaN   
2141900             0.0                     9.0                     NaN   
1310396             0.0                    12.0                    67.0   
792079              0.0                     NaN                     NaN   
1804279             0.0                     NaN                    26.0   

         open_acc  pub_rec  revol_bal  revol_util  total_acc  \
830947       10.0      0.0      301.0         0.3       27.0   
2141900      19.0      0.0    15874.0        49.0       38.0   
1310396       9.0      3.0     4170.0        70.7       25.0   
792079       

In [12]:
df[df.loan_status=='Fully Paid'].groupby('year')['int_rate'].agg('median')

year
2013    13.98
2014    12.99
2015    11.53
2016    11.49
2017    12.62
2018    11.98
Name: int_rate, dtype: float64

In [13]:
set(df.loan_status)

{'Charged Off',
 'Current',
 'Default',
 'Fully Paid',
 'In Grace Period',
 'Late (16-30 days)',
 'Late (31-120 days)'}

In [14]:
df.loan_status.value_counts()

Fully Paid            997912
Current               878317
Charged Off           254245
Late (31-120 days)     21467
In Grace Period         8436
Late (16-30 days)       4349
Default                   40
Name: loan_status, dtype: int64

In [15]:
temp = df[df.loan_status=='Charged Off']
temp.head()

id  loan_amnt  funded_amnt  funded_amnt_inv        term  int_rate  \
13  66624733    18000.0      18000.0          18000.0   60 months     19.48   
25  67849662     4225.0       4225.0           4225.0   36 months     14.85   
30  67715283    16000.0      16000.0          16000.0   36 months     12.88   
31  68341789    24250.0      24250.0          24250.0   60 months     24.24   
33  68415473    25000.0      25000.0          25000.0   60 months     13.99   

    installment grade sub_grade              emp_title emp_length  \
13       471.70     E        E2       Software Manager    7 years   
25       146.16     C        C5               mechanic    5 years   
30       538.18     C        C2  Owner, Fee Appraiser   10+ years   
31       701.01     F        F3        hvac technician    4 years   
33       581.58     C        C4    Executive Assistant    9 years   

   home_ownership  annual_inc verification_status   issue_d  loan_status  \
13           RENT    150000.0        Not Verified  Dec-2015  Charged Off   
25           RENT     35000.0     Source Verified  Dec-2015  Charged Off   
30       MORTGAGE     65000.0        Not Verified  Dec-2015  Charged Off   
31       MORTGAGE     75000.0        Not Verified  Dec-2015  Charged Off   
33       MORTGAGE     79000.0        Not Verified  Dec-2015  Charged Off   

   pymnt_plan                                                url desc  \
13          n  https://lendingclub.com/browse/loanDetail.acti...  NaN   
25          n  https://lendingclub.com/browse/loanDetail.acti...  NaN   
30          n  https://lendingclub.com/browse/loanDetail.acti...  NaN   
31          n  https://lendingclub.com/browse/loanDetail.acti...  NaN   
33          n  https://lendingclub.com/browse/loanDetail.acti...  NaN   

               purpose               title zip_code addr_state    dti  \
13  debt_consolidation  Debt consolidation    916xx         CA   9.39   
25  debt_consolidation  Debt consolidation    672xx         KS  15.22   
30      small_business            Business    354xx         AL  18.96   
31  debt_consolidation                 NaN    114xx         NY  20.84   
33  debt_consolidation  Debt consolidation    212xx         MD  34.53   

    delinq_2yrs earliest_cr_line  fico_range_low  fico_range_high  \
13          0.0         Jun-2005           665.0            669.0   
25          2.0         Jul-2011           725.0            729.0   
30          0.0         Dec-1985           675.0            679.0   
31          0.0         Apr-2007           660.0            664.0   
33          0.0         Jun-2001           730.0            734.0   

    inq_last_6mths  mths_since_last_delinq  mths_since_last_record  open_acc  \
13             1.0                     NaN                    75.0      18.0   
25             0.0                    18.0                     NaN       6.0   
30             0.0                    33.0                    66.0       7.0   
31             0.0                     NaN                     NaN      18.0   
33             1.0                     NaN                     NaN      33.0   

    pub_rec  revol_bal  revol_util  total_acc initial_list_status  out_prncp  \
13      1.0    14052.0        40.7       27.0                   w        0.0   
25      0.0     1058.0        24.6        6.0                   w        0.0   
30      1.0     5157.0        54.3       20.0                   w        0.0   
31      0.0    24799.0        65.3       21.0                   w        0.0   
33      0.0    22519.0        18.6       72.0                   w        0.0   

    out_prncp_inv  total_pymnt  total_pymnt_inv  total_rec_prncp  \
13            0.0      9452.74          9452.74          3481.86   
25            0.0      2558.87          2558.87          1536.90   
30            0.0     17396.63         17396.63         13833.12   
31            0.0      4124.42          4124.42          1332.69   
33            0.0     14490.92         14490.92          8364.55   


In [16]:
temp = temp[['grade','year']]
temp.sample(5)

grade  year
1347390     C  2014
135085      C  2015
1699496     B  2017
98918       B  2015
339376      C  2015

In [17]:
temp.groupby(['grade'])['year'].value_counts()

grade  year
A      2015     3985
       2016     3538
       2014     1947
       2017     1848
       2013      831
       2018      661
B      2016    14334
       2015    13929
       2017     7388
       2014     6813
       2013     4488
       2018     1784
C      2015    23451
       2016    22625
       2017    14401
       2014    12199
       2013     6681
       2018     2751
D      2015    17260
       2016    14617
       2014    10733
       2017     8503
       2013     4600
       2018     2492
E      2015    11957
       2016     8216
       2014     6497
       2017     4197
       2013     2582
       2018      883
F      2015     4194
       2016     3763
       2014     2246
       2017     1730
       2013     1530
       2018      240
G      2016     1149
       2017     1081
       2015     1027
       2014      726
       2013      312
       2018       56
Name: year, dtype: int64

In [18]:
df.groupby('grade')['purpose'].value_counts(normalize=True)

grade  purpose           
A      debt_consolidation    0.477697
       credit_card           0.332728
       home_improvement      0.079246
       other                 0.043052
       major_purchase        0.024550
       car                   0.012553
       medical               0.009856
       house                 0.005714
       small_business        0.005445
       vacation              0.005290
       moving                0.003556
       renewable_energy      0.000313
B      debt_consolidation    0.546258
       credit_card           0.270260
       home_improvement      0.067041
       other                 0.050877
       major_purchase        0.021444
       medical               0.010909
       car                   0.010429
       vacation              0.006356
       small_business        0.006302
       moving                0.004846
       house                 0.004805
       renewable_energy      0.000414
       wedding               0.000057
       educational      

In [19]:
temp = df[df.purpose=='other']
temp.head()

id  loan_amnt  funded_amnt  funded_amnt_inv        term  int_rate  \
12   68338832     1400.0       1400.0           1400.0   36 months     12.88   
37   68407301    27500.0      27500.0          27500.0   60 months     14.85   
49   68387134     2500.0       2500.0           2500.0   36 months     17.27   
101  68416256     5000.0       5000.0           5000.0   36 months      9.80   
116  68356922     4200.0       4200.0           4200.0   36 months     13.99   

     installment grade sub_grade                 emp_title emp_length  \
12         47.10     C        C2         Logistics Manager    3 years   
37        652.06     C        C5                   Manager  10+ years   
49         89.47     D        D3                   Teacher  10+ years   
101       160.87     B        B3  Accounts Payable Manager    6 years   
116       143.53     C        C4            Patrol Officer    2 years   

    home_ownership  annual_inc verification_status   issue_d loan_status  \
12        MORTGAGE     64000.0        Not Verified  Dec-2015  Fully Paid   
37        MORTGAGE    195000.0        Not Verified  Dec-2015  Fully Paid   
49        MORTGAGE     50000.0            Verified  Dec-2015  Fully Paid   
101            OWN     72000.0        Not Verified  Dec-2015  Fully Paid   
116           RENT     32000.0     Source Verified  Dec-2015  Fully Paid   

    pymnt_plan                                                url desc  \
12           n  https://lendingclub.com/browse/loanDetail.acti...  NaN   
37           n  https://lendingclub.com/browse/loanDetail.acti...  NaN   
49           n  https://lendingclub.com/browse/loanDetail.acti...  NaN   
101          n  https://lendingclub.com/browse/loanDetail.acti...  NaN   
116          n  https://lendingclub.com/browse/loanDetail.acti...  NaN   

    purpose  title zip_code addr_state    dti  delinq_2yrs earliest_cr_line  \
12    other    NaN    275xx         NC  34.95          0.0         Jun-1996   
37    other  Other    710xx         LA   6.79          0.0         Jul-2001   
49    other  Other    330xx         FL  17.60          3.0         Jan-2001   
101   other  Other    280xx         NC   6.52          0.0         Jan-2002   
116   other  Other    377xx         TN  31.02          0.0         Nov-2009   

     fico_range_low  fico_range_high  inq_last_6mths  mths_since_last_delinq  \
12            700.0            704.0             0.0                     NaN   
37            700.0            704.0             0.0                     NaN   
49            670.0            674.0             4.0                     6.0   
101           710.0            714.0             0.0                     NaN   
116           705.0            709.0             2.0                    50.0   

     mths_since_last_record  open_acc  pub_rec  revol_bal  revol_util  \
12                      NaN      17.0      0.0    37828.0        67.2   
37                      NaN      14.0      0.0    34974.0        50.9   
49                      NaN       7.0      0.0      732.0        31.1   
101                    89.0      11.0      1.0     4838.0        56.3   
116                     NaN      10.0      0.0     1294.0        11.7   

     total_acc initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
12        24.0                   w        0.0            0.0   1575.160698   
37        19.0                   w        0.0            0.0  29753.170000   
49        22.0                   w        0.0            0.0   3213.299411   
101       20.0                   f        0.0            0.0   5736.458009   
116       16.0                   w        0.0            0.0   5036.372797   

     total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
12           1575.16           1400.0         175.16                 0.0   
37          29753.17          27500.0        2253.17                 0.0   
49           3213.30           2500.0         713.30                 0.0   
101          5736.46 

In [20]:
temp.loan_status.value_counts()

Current               58655
Fully Paid            55638
Charged Off           15183
Late (31-120 days)     1604
In Grace Period         570
Late (16-30 days)       310
Default                   2
Name: loan_status, dtype: int64

In [21]:
#df.groupby('term')['loan_status'].value_counts(normalize=True)
#df.groupby('annual_inc')['loan_status'].value_counts(normalize=True)
# create a list of our conditions
#conditions = [
#    (df['annual_inc'] <= 20000),
#    (df['annual_inc'] > 20000) & (df['annual_inc'] <= 39999),
#    (df['annual_inc'] > 40000) & (df['annual_inc'] <= 59999),
#    (df['annual_inc'] > 60000) & (df['annual_inc'] <= 79999),
#    (df['annual_inc'] > 80000) & (df['annual_inc'] <= 99999),
#    (df['annual_inc'] > 100000) & (df['annual_inc'] <= 119999),
#    (df['annual_inc'] > 120000) & (df['annual_inc'] <= 139999),
#    (df['annual_inc'] > 140000) & (df['annual_inc'] <= 159999),
#    (df['annual_inc'] > 160000) & (df['annual_inc'] <= 179999),
#    (df['annual_inc'] > 180000) & (df['annual_inc'] <= 199999),
#    (df['annual_inc'] > 200000) & (df['annual_inc'] <= 249999),
#    (df['annual_inc'] > 250000)
#    ]

# create a list of the values we want to assign for each condition
#values = ['20k', '40k', '60k', '80k', '100k', '120k', '140k', '160k', '180k', '200k', '250k', '>250k']

# create a new column and use np.select to assign values to it using our lists as arguments
#df['income_bracket'] = np.select(conditions, values)

# display updated DataFrame
#df.head()
#temp = df[df.loan_status=='Default']
#df.groupby('income_bracket')['loan_status'].value_counts(normalize=True)



In [28]:
conditions = [
    (df['term'] == ' 36 months'),
    (df['term'] == ' 60 months')
    ]

# create a list of the values we want to assign for each condition
values = [36, 60]

# create a new column and use np.select to assign values to it using our lists as arguments
df['months'] = np.select(conditions, values)

set(df.emp_length)
#{'1 year',
# '10+ years',
# '2 years',
# '3 years',
# '4 years',
# '5 years',
# '6 years',
# '7 years',
# '8 years',
# '9 years',
# '< 1 year',
# nan}
conditions = [
    (df['emp_length'] == '1 year'),
    (df['emp_length'] == '10+ years'),
    (df['emp_length'] == '2 years'),
    (df['emp_length'] == '3 years'),
    (df['emp_length'] == '4 years'),
    (df['emp_length'] == '5 years'),
    (df['emp_length'] == '6 years'),
    (df['emp_length'] == '7 years'),
    (df['emp_length'] == '8 years'),
    (df['emp_length'] == '9 years'),
    (df['emp_length'] == '< 1 year'),
    (df['emp_length'].isnull().values.any())
    ]

# create a list of the values we want to assign for each condition
values = [1, 10, 2, 3, 4, 5, 6, 7, 8, 9, 0, 0]

# create a new column and use np.select to assign values to it using our lists as arguments
df['emp_years'] = np.select(conditions, values)


set(df.loan_status)
#{'Charged Off',
# 'Current',
# 'Default',
# 'Does not meet the credit policy. Status:Charged Off',
# 'Does not meet the credit policy. Status:Fully Paid',
# 'Fully Paid',
# 'In Grace Period',
# 'Late (16-30 days)',
# 'Late (31-120 days)'}
conditions = [
    (df['loan_status'] == 'Charged Off'),
    (df['loan_status'] == 'Current'),
    (df['loan_status'] == 'Default'),
    (df['loan_status'] == 'Does not meet the credit policy. Status:Charged Off'),
    (df['loan_status'] == 'Does not meet the credit policy. Status:Fully Paid'),
    (df['loan_status'] == 'Fully Paid'),
    (df['loan_status'] == 'In Grace Period'),
    (df['loan_status'] == 'Late (16-30 days)'),
    (df['loan_status'] == 'Late (31-120 days)')
    ]

# create a list of the values we want to assign for each condition
values = ['Delinquent', 'Good', 'Delinquent', 'Delinquent', 'Good', 'Good', 'Delinquent', 'Delinquent', 'Delinquent']

# create a new column and use np.select to assign values to it using our lists as arguments
df['loan_status2'] = np.select(conditions, values)

df.head()

df["income"] = df[["annual_inc_joint", "annual_inc"]].max(axis=1)
df["dti2"] = df[["dti", "dti_joint"]].min(axis=1)
df = df[df['earliest_cr_line'].notna()]
df['early_credit_year']=df.earliest_cr_line.map(lambda x:int(x.split('-')[-1]))
df = df[df['zip_code'].notna()]
df['emp_title'] = df['emp_title'].fillna("none")
df['title'] = df['title'].fillna("none")
df['inq_last_6mths'] = df['inq_last_6mths'].fillna(0)
df['mths_since_last_delinq'] = df['mths_since_last_delinq'].fillna(85)
df['mths_since_last_record'] = df['mths_since_last_record'].fillna(85)
df['revol_util'] = df['revol_util'].fillna(0)
df['mths_since_last_major_derog'] = df['mths_since_last_major_derog'].fillna(85)
df['open_acc_6m'] = df['open_acc_6m'].fillna(0)
df['open_act_il'] = df['open_act_il'].fillna(0)
df['open_il_12m'] = df['open_il_12m'].fillna(0)
df['open_il_24m'] = df['open_il_24m'].fillna(0)
df['mths_since_rcnt_il'] = df['mths_since_rcnt_il'].fillna(85)
df['il_util'] = df['il_util'].fillna(0)
df['open_rv_12m'] = df['open_rv_12m'].fillna(0)
df['open_rv_24m'] = df['open_rv_24m'].fillna(0)
df['max_bal_bc'] = df['max_bal_bc'].fillna(0)
df['all_util'] = df['all_util'].fillna(0)
df['inq_fi'] = df['inq_fi'].fillna(0)
df['total_cu_tl'] = df['total_cu_tl'].fillna(0)
df['inq_last_12m'] = df['inq_last_12m'].fillna(0)
df['avg_cur_bal'] = df['avg_cur_bal'].fillna(0)
df['bc_open_to_buy'] = df['bc_open_to_buy'].fillna(0)
df['bc_util'] = df['bc_util'].fillna(0)
df['mo_sin_old_il_acct'] = df['mo_sin_old_il_acct'].fillna(85)
df['mths_since_recent_bc'] = df['mths_since_recent_bc'].fillna(85)
df['mths_since_recent_bc_dlq'] = df['mths_since_recent_bc_dlq'].fillna(85)
df['mths_since_recent_inq'] = df['mths_since_recent_inq'].fillna(85)
df['mths_since_recent_revol_delinq'] = df['mths_since_recent_revol_delinq'].fillna(85)
df['num_tl_120dpd_2m'] = df['num_tl_120dpd_2m'].fillna(0)
df['num_rev_accts'] = df['num_rev_accts'].fillna(0)
df['percent_bc_gt_75'] = df['percent_bc_gt_75'].fillna(0)
df['pct_tl_nvr_dlq'] = df['pct_tl_nvr_dlq'].fillna(100)
df["revol_bal2"] = df[["revol_bal_joint", "revol_bal"]].max(axis=1)
df['inq_last_6mths2'] = df[['sec_app_inq_last_6mths','inq_last_6mths']].sum(axis=1)
df['mort_acc2'] = df[['sec_app_mort_acc','mort_acc']].sum(axis=1)
df['open_act_il2'] = df[['sec_app_open_act_il','open_act_il']].sum(axis=1)
df['num_rev_accts2'] = df[['sec_app_num_rev_accts','num_rev_accts']].sum(axis=1)
df['chargeoff_within_12_mths2'] = df[['sec_app_chargeoff_within_12_mths','chargeoff_within_12_mths']].sum(axis=1)
df['collections_12_mths_ex_med2'] = df[['sec_app_collections_12_mths_ex_med','collections_12_mths_ex_med']].sum(axis=1)
df['sec_app_mths_since_last_major_derog'] = df['sec_app_mths_since_last_major_derog'].fillna(85)
df['mths_since_last_major_derog2'] = df[['sec_app_mths_since_last_major_derog','mths_since_last_major_derog']].min(axis=1)
df['sec_app_earliest_cr_line'] = df['sec_app_earliest_cr_line'].fillna("October-2020")
df['sec_app_early_credit_year']=df.sec_app_earliest_cr_line.map(lambda x:int(x.split('-')[-1]))
df['early_credit_year2'] = df[['sec_app_early_credit_year','early_credit_year']].min(axis=1)
df['open_acc2']=df[['sec_app_open_acc','open_acc']].sum(axis=1)

#sec_app_earliest_cr_line               2056744

#df = df[df['sec_app_earliest_cr_line'].notna()]
#df['sec_early_credit_year']=df.sec_app_earliest_cr_line.map(lambda x:int(x.split('-')[-1]))
df['fico_low']=df[["sec_app_fico_range_low","fico_range_low"]].max(axis=1)
df['fico_high']=df[["sec_app_fico_range_high","fico_range_high"]].max(axis=1)

#df.drop(columns=['pymnt_plan', 'url', 'desc', 'delinq_2yrs', 'pub_rec', 'revol_util', 'initial_list_status', 'out_prncp', 'out_prncp_inv', 'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int', 'total_rec_late_fee', 'recoveries', 'collection_recovery_fee', 'last_pymnt_d', 'last_pymnt_amnt', 'next_pymnt_d', 'last_credit_pull_d', 'last_fico_range_high', 'last_fico_range_low', 'collections_12_mths_ex_med', 'mths_since_last_major_derog', 'policy_code', 'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m', 'open_act_il', 'open_il_12m', 'open_il_24m', 'mths_since_rcnt_il', 'total_bal_il', 'il_util', 'open_rv_12m', 'open_rv_24m', 'max_bal_bc', 'all_util', 'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util', 'chargeoff_within_12_mths', 'delinq_amnt', 'mo_sin_old_il_acct', 'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl', 'mort_acc', 'mths_since_recent_bc', 'mths_since_recent_bc_dlq', 'mths_since_recent_inq', 'mths_since_recent_revol_delinq', 'num_accts_ever_120_pd', 'num_actv_bc_tl', 'num_actv_rev_tl', 'num_bc_sats', 'num_bc_tl', 'num_il_tl', 'num_op_rev_tl', 'num_rev_accts', 'num_rev_tl_bal_gt_0', 'num_sats', 'num_tl_120dpd_2m', 'num_tl_30dpd', 'num_tl_90g_dpd_24m', 'num_tl_op_past_12m', 'hardship_flag', 'hardship_type', 'hardship_reason', 'hardship_status', 'deferral_term', 'hardship_amount', 'hardship_start_date', 'hardship_end_date', 'payment_plan_start_date', 'hardship_length', 'hardship_dpd', 'hardship_loan_status', 'orig_projected_additional_accrued_interest', 'hardship_payoff_balance_amount', 'hardship_last_payment_amount', 'debt_settlement_flag', 'debt_settlement_flag_date', 'settlement_status', 'settlement_date', 'settlement_amount', 'settlement_percentage', 'settlement_term'])
df2 = df.drop(columns=['id','sec_app_open_acc','open_acc','sec_app_earliest_cr_line','sec_app_early_credit_year','early_credit_year','sec_app_collections_12_mths_ex_med','collections_12_mths_ex_med','sec_app_chargeoff_within_12_mths', 'chargeoff_within_12_mths','sec_app_num_rev_accts','num_rev_accts','sec_app_open_act_il','open_act_il','sec_app_revol_util','sec_app_mort_acc','mort_acc','sec_app_inq_last_6mths','inq_last_6mths','revol_bal','revol_bal_joint','chargeoff_within_12_mths','delinq_amnt','desc','sec_app_fico_range_low','fico_range_low','sec_app_fico_range_high','fico_range_high','earliest_cr_line','term','dti', 'dti_joint','annual_inc', 'annual_inc_joint','verification_status_joint','acc_now_delinq', 'tot_coll_amt','tot_cur_bal','total_bal_il','out_prncp', 'out_prncp_inv', 'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int', 'total_rec_late_fee', 'recoveries', 'collection_recovery_fee', 'last_pymnt_d', 'last_pymnt_amnt', 'next_pymnt_d', 'last_credit_pull_d', 'last_fico_range_high', 'last_fico_range_low', 'policy_code','pymnt_plan', 'url', 'initial_list_status', 'emp_length', 'disbursement_method','hardship_flag', 'hardship_type', 'hardship_reason', 'hardship_status', 'deferral_term', 'hardship_amount', 'hardship_start_date', 'hardship_end_date', 'payment_plan_start_date', 'hardship_length', 'hardship_dpd', 'hardship_loan_status', 'orig_projected_additional_accrued_interest', 'hardship_payoff_balance_amount', 'hardship_last_payment_amount', 'debt_settlement_flag', 'debt_settlement_flag_date', 'settlement_status', 'settlement_date', 'settlement_amount', 'settlement_percentage', 'settlement_term'])

df2.head()

#dti_joint
#sec_app_earliest_cr_line	sec_app_inq_last_6mths	sec_app_mort_acc	sec_app_open_acc	sec_app_revol_util	sec_app_open_act_il	sec_app_num_rev_accts	sec_app_chargeoff_within_12_mths	sec_app_collections_12_mths_ex_med	sec_app_mths_since_last_major_derog	


loan_amnt  funded_amnt  funded_amnt_inv  int_rate  installment grade  \
0     3600.0       3600.0           3600.0     13.99       123.03     C   
1    24700.0      24700.0          24700.0     11.99       820.28     C   
2    20000.0      20000.0          20000.0     10.78       432.66     B   
3    35000.0      35000.0          35000.0     14.85       829.90     C   
4    10400.0      10400.0          10400.0     22.45       289.91     F   

  sub_grade                    emp_title home_ownership verification_status  \
0        C4                      leadman       MORTGAGE        Not Verified   
1        C1                     Engineer       MORTGAGE        Not Verified   
2        B4                 truck driver       MORTGAGE        Not Verified   
3        C5  Information Systems Officer       MORTGAGE     Source Verified   
4        F1          Contract Specialist       MORTGAGE     Source Verified   

    issue_d loan_status             purpose               title zip_code  \
0  Dec-2015  Fully Paid  debt_consolidation  Debt consolidation    190xx   
1  Dec-2015  Fully Paid      small_business            Business    577xx   
2  Dec-2015  Fully Paid    home_improvement                none    605xx   
3  Dec-2015     Current  debt_consolidation  Debt consolidation    076xx   
4  Dec-2015  Fully Paid      major_purchase      Major purchase    174xx   

  addr_state  delinq_2yrs  mths_since_last_delinq  mths_since_last_record  \
0         PA          0.0                    30.0                    85.0   
1         SD          1.0                     6.0                    85.0   
2         IL          0.0                    85.0                    85.0   
3         NJ          0.0                    85.0                    85.0   
4         PA          1.0                    12.0                    85.0   

   pub_rec  revol_util  total_acc  mths_since_last_major_derog  \
0      0.0        29.7       13.0                         30.0   
1      0.0        19.2       38.0                         85.0   
2      0.0        56.2       18.0                         85.0   
3      0.0        11.6       17.0                         85.0   
4      0.0        64.5       35.0                         85.0   

  application_type  open_acc_6m  open_il_12m  open_il_24m  mths_since_rcnt_il  \
0       Individual          2.0          0.0          1.0                21.0   
1       Individual          1.0          0.0          1.0                19.0   
2        Joint App          0.0          0.0          4.0                19.0   
3       Individual          1.0          0.0          1.0                23.0   
4       Individual          1.0          0.0          3.0                14.0   

   il_util  open_rv_12m  open_rv_24m  max_bal_bc  all_util  total_rev_hi_lim  \
0     36.0          3.0          3.0       722.0      34.0            9300.0   
1     73.0          2.0          3.0      6472.0      29.0          111800.0   
2     73.0          0.0          2.0      2081.0      65.0           14000.0   
3     70.0          1.0          1.0      6987.0      45.0           67300.0   
4     84.0          4.0          7.0      9702.0      78.0           34000.0   

   inq_fi  total_cu_tl  inq_last_12m  acc_open_past_24mths  avg_cur_bal  \
0     3.0          1.0           4.0                   4.0      20701.0   
1     0.0          0.0           6.0                   4.0       9733.0   
2     2.0          5.0           1.0                   6.0      31617.0   
3     0.0          1.0           0.0                   2.0      23192.0   
4     2.0          1.0           3.0                  10.0      27644.0   

   bc_open_to_buy  bc_util  mo_sin_old_il_acct  mo_sin_old_rev_tl_op  \
0          1506.0     37.2               148.0                 128.0   
1         57830.0     27.1               113.0                 192.0   
2          2737.0     55.9               125.0                 184.0   
3         54962.0     12.1                36.0       

In [29]:
df2.isnull().sum()


loan_amnt                              0
funded_amnt                            0
funded_amnt_inv                        0
int_rate                               0
installment                            0
grade                                  0
sub_grade                              0
emp_title                              0
home_ownership                         0
verification_status                    0
issue_d                                0
loan_status                            0
purpose                                0
title                                  0
zip_code                               0
addr_state                             0
delinq_2yrs                            0
mths_since_last_delinq                 0
mths_since_last_record                 0
pub_rec                                0
revol_util                             0
total_acc                              0
mths_since_last_major_derog            0
application_type                       0
open_acc_6m     

In [30]:
df2.shape

(2164765, 89)

In [31]:
df2.to_csv('revised_accepted_2013_to_2018Q4.csv', index=False)

In [25]:
temp.loan_status.value_counts()

Current               58655
Fully Paid            55638
Charged Off           15183
Late (31-120 days)     1604
In Grace Period         570
Late (16-30 days)       310
Default                   2
Name: loan_status, dtype: int64

In [26]:
# average return for all loan that has a fico score > 800
total = np.sum(temp.loan_amnt)
temp['weight'] = temp.loan_amnt/total
temp['weight'] = temp['weight'] * temp.int_rate
np.sum(temp['weight'])

14.536408587799079

 #  1 USD in SP500 vs Lending club (Fico >800) FROM 2007 to 2018 (12 Year)
 # 8.3 vs 7.26

In [27]:
# fv=pv(1+i)^n


- Installment Credit: predetermined length and end date such as mortgage, car loan
- Revolving Credit: No payment plan  higher interest rates such as credit card loan

- Amortizing loan:Amortization is the process of spreading out a loan into a series of fixed payments. The loan is paid off at the end of the payment schedule

- loan default: A loan goes into default—which is the eventual consequence of extended payment delinquency
- loan delinquent: Payment delinquency is commonly used to describe a situation in which a borrower misses a single payment

- loan settlement is the act of paying back the amount of money owed to the lender
- loan collection is the process of pursuing payments of debts owed by individuals or businesses

- 'Fully Paid', 'Current', 'Late', 'Grace period', 'Default', 'Charged Off'
- late to default  before charged off
- default: a note goes into Default status when it is 121 or more days past due.
- Charge-Off:When a note is in Default status, Charge Off occurs no later than 150 days past due (i.e. No later than 30 days after the Default status is reached) when there is no reasonable expectation of sufficient payment to prevent the charge off. 